crates/ruff_linter/src/rules/refurb/rules/for_loop_writes.rs

```diff
@@ -113,16 +182,55 @@ pub(crate) fn for_loop_writes(checker: &mut Checker, for_stmt: &StmtFor) {
         Applicability::Safe
     };
 
-    checker.diagnostics.push(
-        Diagnostic::new(
-            ForLoopWrites {
-                name: io_object_name.id.to_string(),
-            },
-            for_stmt.range,
-        )
-        .with_fix(Fix::applicable_edit(
-            Edit::range_replacement(content, for_stmt.range),
-            applicability,
-        )),
+    let diagnostic = Diagnostic::new(
+        ForLoopWrites {
+            name: io_object_name.id.to_string(),
+        },
+        for_stmt.range,
     );
+    let fix = Fix::applicable_edit(
+        Edit::range_replacement(content, for_stmt.range),
+        applicability,
+    );
+
+    Some(diagnostic.with_fix(fix))
+}
+
+fn loop_variables_are_used_outside_loop(
+    binding_names: &[&ExprName],
+    loop_range: TextRange,
+    semantic: &SemanticModel,
+    scope_id: ScopeId,
+) -> bool {
+    let find_binding_id = |name: &ExprName, offset: TextSize| {
+        semantic.simulate_runtime_load_at_location_in_scope(
+            name.id.as_str(),
+            TextRange::at(offset, 0.into()),
+            scope_id,
+            TypingOnlyBindingsStatus::Disallowed,
+        )
+    };
+
+    let name_overwrites_outer =
+        |name: &ExprName| find_binding_id(name, loop_range.start()).is_some();
```

Comment:
Could you add some comments here that explain what these steps do. I think I understand what's going on, but it's not trivial to understand for a new reader of the code